syntax = "proto3";

package signals;

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ============================================================================
// Service Definition
// ============================================================================

service CryptoStreamingService {
  // Main bidirectional streaming RPC for crypto providers
  rpc ConnectCryptoProvider(stream SignalProviderMessage)
      returns (stream DjangoCommand);

  // Cross-process command routing (for management commands)
  rpc SendCommandToProvider(SendCommandRequest)
      returns (SendCommandResponse);

  // Synchronous command execution with response (like trading_bots)
  rpc ExecuteCommandSync(SendCommandRequest)
      returns (CommandAck);
}

// ============================================================================
// Provider → Django Messages
// ============================================================================

message SignalProviderMessage {
  string provider_id = 1;              // UUID of provider (Source.id)
  string message_id = 2;               // Tracking ID
  google.protobuf.Timestamp timestamp = 3;

  oneof payload {
    RegisterProviderRequest register = 10;
    SignalReport signal_report = 11;   // Main payload: trading signal
    HeartbeatUpdate heartbeat = 12;
    StatusUpdate status = 13;
    CommandAck command_ack = 14;       // Command acknowledgment
    ErrorReport error = 15;
    LogEntry log = 16;
  }
}

// Provider registration on connection
message RegisterProviderRequest {
  string provider_name = 1;
  ProviderType provider_type = 2;
  string version = 3;
  string url = 4;                       // Optional
  string description = 5;
  google.protobuf.Struct metadata = 6;  // Additional data (JSON)
}

// Trading signal from provider
message SignalReport {
  string signal_id = 1;                    // Optional UUID (auto-generated if not provided)
  SignalType signal_type = 2;              // BUY/SELL/INFO
  int32 priority = 3;                      // 0-100 (higher = more important)
  string content = 4;                      // Human-readable signal description
  google.protobuf.Timestamp signal_date = 5;

  // Coin target (ONE required: either target_coin_symbol OR coin_pair)
  optional string target_coin_symbol = 10;  // Single coin: "BTC"
  optional CoinPair coin_pair = 11;         // Trading pair: BTC/USDT

  // Metadata
  google.protobuf.Struct raw_data = 20;     // Additional JSON data
  repeated string tags = 21;                // Tags for categorization
  optional double confidence = 22;          // Confidence level 0.0-1.0
}

// Trading pair definition
message CoinPair {
  string base_symbol = 1;               // "BTC"
  string quote_symbol = 2;              // "USDT"
}

// Heartbeat - empty message, keepalive only (no data transfer)
message HeartbeatUpdate {
  // Empty - just keepalive signal
  // Universal pattern: heartbeat presence = connection alive
  // No metrics, no status - just technical keepalive
}

// Provider status change notification
message StatusUpdate {
  ProviderStatus status = 1;
  string message = 2;
}

// Error notification from provider
message ErrorReport {
  ErrorLevel level = 1;
  string error_code = 2;
  string message = 3;
  google.protobuf.Struct context = 4;
}

// Log message from provider
message LogEntry {
  LogLevel level = 1;
  string message = 2;
  google.protobuf.Struct data = 3;
}

// Command acknowledgment from provider
message CommandAck {
  string command_id = 1;              // Original command ID
  bool success = 2;                   // true = executed, false = failed
  string message = 3;                 // Status message
  ProviderStatus current_status = 4;  // Provider status after command
  optional string error = 5;          // Error details if failed
}

// ============================================================================
// Django → Provider Commands
// ============================================================================

message DjangoCommand {
  string command_id = 1;                // UUID
  google.protobuf.Timestamp timestamp = 2;

  oneof payload {
    AckSignalCommand ack_signal = 10;       // Signal acknowledgment
    RequestStatusCommand request_status = 11;
    PauseProviderCommand pause = 12;
    ResumeProviderCommand resume = 13;
    PingCommand ping = 14;
  }
}

// Signal acknowledgment from Django
message AckSignalCommand {
  string signal_id = 1;                  // Original signal ID
  bool accepted = 2;                     // true = saved, false = rejected
  optional string rejection_reason = 3;  // Reason if rejected
  optional string created_signal_id = 4; // UUID in Django DB if accepted
}

// Pause signal sending
message PauseProviderCommand {
  string reason = 1;
}

// Resume signal sending
message ResumeProviderCommand {
  string message = 1;
}

// Request current status from provider
message RequestStatusCommand {
  bool include_stats = 1;
}

// Keepalive ping (sent every 30s if no other commands)
message PingCommand {
  int32 sequence = 1;
}

// ============================================================================
// Cross-process Communication
// ============================================================================

message SendCommandRequest {
  string provider_id = 1;
  DjangoCommand command = 2;
}

message SendCommandResponse {
  bool success = 1;
  optional string error = 2;
}

// ============================================================================
// Enums
// ============================================================================

enum ProviderType {
  PROVIDER_TYPE_UNSPECIFIED = 0;
  PROVIDER_TELEGRAM = 1;
  PROVIDER_API = 2;
  PROVIDER_WEBHOOK = 3;
  PROVIDER_BOT = 4;
  PROVIDER_MANUAL = 5;
}

enum SignalType {
  SIGNAL_TYPE_UNSPECIFIED = 0;
  SIGNAL_BUY = 1;
  SIGNAL_SELL = 2;
  SIGNAL_INFO = 3;
}

enum ProviderStatus {
  PROVIDER_STATUS_UNSPECIFIED = 0;
  PROVIDER_ACTIVE = 1;
  PROVIDER_PAUSED = 2;
  PROVIDER_ERROR = 3;
}

enum ErrorLevel {
  ERROR_LEVEL_UNSPECIFIED = 0;
  ERROR_WARNING = 1;
  ERROR_ERROR = 2;
  ERROR_CRITICAL = 3;
}

enum LogLevel {
  LOG_LEVEL_UNSPECIFIED = 0;
  LOG_DEBUG = 1;
  LOG_INFO = 2;
  LOG_WARNING = 3;
  LOG_ERROR = 4;
}
